inc/class-db.php

```diff
@@ -33,6 +76,39 @@ function query( $query ) {
 			XRay\trace_wpdb_query( $query, $start, $end, $result === false ? $this->last_error : null, $host );
 		}
 		$this->time_spent += $end - $start;
+
+		if ( ! $has_qm || ! SAVEQUERIES ) {
+			return $result;
+		}
+
+		$i = count( $this->queries ) - 1;
+		$this->queries[ $i ]['trace'] = new QM_Backtrace( [
+			'ignore_frames' => 1,
+		] );
+
+		if ( ! isset( $this->queries[ $i ][3] ) ) {
+			$this->queries[ $i ][3] = $this->time_start;
+		}
+
+		if ( $this->last_error ) {
+			$code = 'qmdb';
+			if ( $this->use_mysqli ) {
+				if ( $this->dbh instanceof mysqli ) {
+					// phpcs:ignore WordPress.DB.RestrictedFunctions.mysql_mysqli_errno
+					$code = mysqli_errno( $this->dbh );
+				}
+			} else {
+				if ( is_resource( $this->dbh ) ) {
+					// Please do not report this code as a PHP 7 incompatibility. Observe the surrounding logic.
+					// @codingStandardsIgnoreLine
+					$code = mysql_errno( $this->dbh );
+				}
+			}
+			$this->queries[ $i ]['result'] = new WP_Error( $code, $this->last_error );
+		} else {
+			$this->queries[ $i ]['result'] = $result;
+		}
+
```

Comment:
I'll see if I can abstract this out into a separate function in QM.

Reply:
Thanks!